### ISSUE TITLE
Ability to aggregate over joined facts

### ISSUE THREAD

**Issue by snikolayev** [NRules/NRules#104]
Rule engine is currently capable of joining facts (i.e. matching facts correlated via a join condition) as well as aggregating facts on the fly (i.e. collecting matching facts into a collection).
What's currently very hard to do is to join facts and then aggregate over those joined fact.
I.e., in pseudo code:
Match a
Match b (joined to a)
Collect all matched (a,b) tuples as IEnumerable

**Comment by DonaldAirey** [comment]
Yes, this is very important to a state-driven rules engine.  Ideally, you would want a single alert to be generated when a violation of a given rules occurs, and you would want that same alert to be cleared when the condition goes away.  In order to do this, each rule evaluation must be associated with a single .Do action (and a unique key so we have a way of clearing the violation when the **.Undo** action is triggered).  Aggregates (GroupBy) rules invoke a different **.Do** for every condition in the **IGrouping** so it's impossible to know when a condition has been cleared (because they are ambiguous without a unique key).  Being able to roll the facts back up to a single rule violation is critical.
